src/psyclone/dynamo0p3.py

```diff
@@ -2896,29 +2900,30 @@ def __init__(self, node):
         # Create lists of real and integer scalar arguments
         self._real_scalar_names = {}
         self._int_scalar_names = {}
+        for intent in FORTRAN_INTENT_NAMES:
```

Comment:
This implementation now contains a mix of the old (assumes just 'real' and 'int') and the new (allows for all VALID_SCALAR_NAMES). Could we instead just use the dict of scalar arguments throughout the class, filtering as required? Alternatively, we could have a new `_scalar_names` dict with each entry indexed by intent and itself consisting of a dict indexed by intrinsic type?
(Presumably you're going to change VALID_SCALAR_NAMES to just be "gh_scalar" at some point (future PR) with the type information specified separately?)

Reply:
Yes, I thought about that and I think there are two potential solutions:
1. Modify  `unique_declns_by_intent` and `unique_declarations` to accept a list of names instead of just one name (this would cover as for the change in number of argument names) and do filterinb by intrinsic type in the API;
2. 1 + modify these two functions to also do filtering by intrinsic type .

I am happy to do either, as they are only used in the LFRic API for now.

Reply:
I went for the first solution, and that is to modify `unique_declns_by_intent` and `unique_declarations` to accept a list of names instead of just one name (also did it for `unique_proxy_declarations` for consistency).

The `DynScalarArgs.__init__` method is now rewritten to first return a dictionary of scalar arguments by intent for both Invoke and Kernel objects. Then the arguments are filtered by intrinsic types to populate the dictionaries for `real` and `integer` scalars.

As noted from a comment above, the invalid intrinsic type is now caught when setting up `DynKernelArgument.intrinsic_type` property (creates intrinsic type from `dynamo0p3.MAPPING_DATA_TYPES` and checks the mapping there).